extensions/ql-vscode/src/common/helpers-pure.ts

```diff
@@ -7,6 +7,18 @@
 
 import { RedactableError } from "./errors";
 
+// Matches any type that is not an array. This is useful to help avoid
+// nested arrays, or for cases like createSingleSelectionCommand to avoid T
+// accidentally getting instantiated as DatabaseItem[] instead of DatabaseItem.
+export type NotArray =
```

Comment:
How about this insted?
```ts
type NonArray<T> = T extends any[] ? never : T;
```

Reply:
Unfortunately I don't think this works, unless I've made a mistake in my testing.

Note I had to make a couple of modifications:
- We don't want to use `any`, so I changed that to `T extends unknown[]`
- Having `NotArray` have a type argument itself is a little awkward. Using it could force us to do something like `foo<T extends NotArray<S>, S>` which is weird because `foo` now has an extra type arg it doesn't really need. Instead I went with `NotArray<T = unknown` to provide a default so it can be used just as `NotArray`.

Sadly it doesn't appear to be working. I can do `const x: NotArray = [1, 2, 3];` and it doesn't cause any errors, while it does cause an error with the `NotArray` definition in this PR.

Reply:
Hmm, I now see that `NotArray<T = unknown>` basically means the ternary compiles down to `T = unknown`, so that isn't very helpful. That explains why it wasn't causing an error in my test.

Unfortunately I still don't see how it make it work as intended with this approach.

@charisk, do you have any explicit issues with the current approach in this PR that should be changed?

Reply:
Ah okay, not a great suggestion then. I was just not super keen on having to define every type that isn't an array, but it's not bad.

Reply:
At least it's not a very long list, and it isn't dependent on anything we do in our code so it shouldn't ever have to change.